text/0000-js-compatible-resolution.md

```diff
@@ -0,0 +1,344 @@
+
+```js
+//config/template-globals.js
+export default {
+  t: "import { t } from 'ember-intl'"
+};
+```
+
+This says "whenever the template compiler encounters the undefined name `t`, it should automatically insert `"import { t } from 'ember-intl'"` at the start of the template. This only applies to otherwise _undefined_ names: if a template creates its own binding for `t`, that takes precedence.
+
+Addons are deliberately _not_ allowed to register themselves into the template-globals. They can offer a generator that helps app-authors edit `template-globals.js`, but control over globals remains _explicit in the app_, so that there's only one place where you need to look to find where a global name is coming from.
+
+## Rationalizing Components as Modules
+
+A key piece of this design is explaining how to interpret Ember components as Javascript modules. There are some important requirements:
+
+ - When a component has a Javascript module, and you import the component from Javascript, you should get the values back you'd expect for a normal Javascript module. We don't want to rewrite things in an opaque way.
+ - Consumers of a component should not need to know whether it is implemented as only a template, a template plus Javascript, or only Javascript.
+ - A component that is reexported should continue to work unchanged (in contrast with today, in which component authors needs to take special care to make their templates come along).
+ - We want to follow the _existing_ Module Resolver rules and not introduce new ones. Especially new ones that would diverge further from typical node_modules resolution.
+
+Given these requirements, the proposed design is:
+
+1. Imports in templates are resolved the same way they are resolved today in Javascript. Conceptually, we add a custom handler for `.hbs` files that allows them to be resolved the same as `.js` files. `.js` files take precedence over `.hbs` files, such that if both exist for a given ModuleSpecifier, the JS wins.
+
+2. Component Javascript should be authored using default exports (as they already are):
+
+    ```js
+    export default class extends Component {
+      yourStuffHere(){}
+    };
+    ```
+
+    There is no magic that alters the exported value. When it's imported (whether in Javacript or HBS), the value you get is the actual class.
+
+3. When both `.js` and `.hbs` exist for a component, the build system unobtrusively associates the template with the default value exported by the Javascript. The template must be co-located with the Javascript, having the same filename except for changing the extension to `.hbs`.
+
+    For illustrative purposes, here's a toy implementation of how we can do "unobtrusive" association of the template with the Javascript value:
+
+    ```js
+    // the previous example code would get rewritten to something like this:
+    import { registerTemplate } from '@ember/hypothetical-private-api';
```

Comment:
FWIW, I do not think this can actually be private API. I totally agree that this is a tooling API and end users do not need to deal with it or understand it. 

I'm suggesting we lint against its direct usage...

Reply:
It would be really nice to a) expose this API to addon authors and b) keep it generic, so that it may be used for different file types as well.

This is of particular interest for addons like [ember-css-modules](https://github.com/salsify/ember-css-modules) that currently have to run an hbs transform that [injects the file path of the template into the template helpers](https://github.com/salsify/ember-css-modules/blob/dfc795df486d26dc8c3b15b9b68823fd3cbe2c90/lib/htmlbars-plugin/index.js#L69-L74), so that they can be associated with the co-located `.css` files.